File: 2018/day2/part1.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

int main()
{
	std::string line;
	std::vector<std::string> vec;
	while (std::getline(std::cin, line)) vec.push_back(line);
	unsigned two_count = 0, three_count = 0;
	for (const auto &word : vec) {
		std::map<char, int> letter_counts;
		for (const char &ch : word) {
			letter_counts[ch]++;
		}
		bool found_two = false, found_three = false;
		for (const auto &pair : letter_counts) {
			if (pair.second == 2 && !found_two) {
				found_two = true;
				two_count++;
			}
			else if (pair.second == 3 && !found_three) {
				found_three = true;
				three_count++;
			}
		}
	}
	std::cout << two_count * three_count << '\n';
	return 0;
}

File: 2018/day2/part1.swift
import Foundation

func makeLetterCounts(_ word: String) -> [Character: Int] {
    var ret = [Character: Int]()
    for ch in word {
        if let chCount = ret[ch] {
            ret[ch] = chCount + 1
        } else {
            ret[ch] = 1
        }
    }
    return ret
}

func part1(_ input: String) -> Int {
    let inputSplit = input.components(separatedBy: "\n").map { makeLetterCounts($0) }
    var twoCount = 0, threeCount = 0
    for dict in inputSplit {
        var foundTwo = false, foundThree = false
        for (_, count) in dict {
            if count == 2 && !foundTwo {
                twoCount += 1
                foundTwo = true
            } else if count == 3 && !foundThree {
                threeCount += 1
                foundThree = true
            }
        }
    }
    return twoCount * threeCount
}
